tools/releaser/lib/generate_changelog.dart

```diff
@@ -0,0 +1,155 @@
+    if (!file.existsSync()) {
+      Logger.root.shout('❌ Could not find file CHANGELOG.md for package.');
+      return false;
+    }
+
+    bool didWriteChangelog = false;
+    await transformFile(file, logger, args.dryRun, (line) {
+      if (didWriteChangelog) return line;
+
+      if (line.startsWith('##')) {
+        String? oldLine = line;
+        if (line == '## Unreleased') {
+          logger
+              .info('ℹ️ ## Unreleased headers are no longer needed. Removing.');
+          oldLine = null;
+        }
+
+        line = '## ${args.version}\n\n$versionChangelog';
+        if (oldLine != null) {
+          line += '\n\n$oldLine';
+        }
+        didWriteChangelog = true;
+      }
+      return line;
+    });
+
+    print(
+        'Verify the CHANGELOG.md changes and add changes from iOS and Android Native SDK updates.');
+    print(
+        'For reference iOS SDK will be updated to ${args.iOSRelease} and Android SDK will be updated to ${args.androidRelease}.');
+    print('Ready to continue? ([Y]es, [N]o): ');
+
+    final input = stdin.readLineSync();
+    if (input != null && input.isNotEmpty) {
+      final firstChar = input[0].toLowerCase();
+      if (firstChar == 'y') {
+        return true;
+      } else if (firstChar == 'n') {
+        logger.shout('😳 Oh, I\'m glad we stopped then!');
+        return false;
+      } else {
+        logger.shout(
+            '❓ Not sure what you meant by that... stopping just in case.');
+        return false;
+      }
+    }
+
+    return true;
+  }
+
+  Future<String> _findLastReleaseSha(CommandArguments args) async {
+    final packageTags = await args.gitDir
+        .tags()
+        .where((t) => t.tag.startsWith('${args.packageName}/'))
+        .toList();
+
+    return packageTags.last.objectSha;
+  }
+
+  Future<List<String>> _getCommits(
+      CommandArguments args, String commitRange) async {
+    final result = await args.gitDir.runCommand([
+      '--no-pager',
+      'log',
+      commitRange,
+      '--pretty=format:%H|||%an <%aE>|||%ai|||%B||||',
+      '--',
+      args.packageRoot
+    ]);
+
+    final rawCommits = (result.stdout as String)
+        .split('||||\n')
+        .where((e) => e.trim().isNotEmpty)
+        .toList();
+
+    return rawCommits.map((c) {
+      final parts = c.split('|||');
+      return parts[3].trim();
+    }).toList();
+  }
+}
+
+List<String> _getChangelogItems(List<String> commitMessages) {
+  RegExp conventionalCommitPattern =
+      RegExp(r'(?<type>.*)(\((?<scope>.*)\))?(?<breaking>!)?: (?<rest>.*)');
+  RegExp githubIssueMention = RegExp(r'\#(?<issue_number>\d+)');
+
+  final items = <String>[];
+  for (final commitMessage in commitMessages) {
+    final lines = commitMessage.split('\n');
+    final summaryLine = lines[0];
+    final match = conventionalCommitPattern.firstMatch(summaryLine);
+    if (match != null) {
+      final type = match.namedGroup('type');
+      if (type != null && (type == 'fix' || type == 'feat')) {
```

Comment:
very minor: is `type != null` check is needed here? I think check `type == 'fix' || type == 'feat'` is enough.

Reply:
Good catch.